File: leetcode/medium/151.Reverse Words in a String/python/reverseWordsInAString.py
"""Reverse Words in a String"""

def reverse_words(s):
    """Return a string of the words in reverse order concatenated by a single space."""
    reverse_array = []
    begin_word = 0
    end_word = 0
    while end_word < len(s):
        word = ""
        while s[end_word] != " " and end_word < len(s):
            end_word += 1
            if end_word == len(s):
                break
        if begin_word < end_word:
            word = s[begin_word : end_word]
            reverse_array.insert(0, word)
            begin_word = end_word
        begin_word += 1
        end_word += 1
    s = " ".join(reverse_array)
    return s

# Testing 
s0 = " the sky is blue "
s1 = "  hello world  "
s2 = "a good   example"
s3 = " asdasd df f"
print(reverse_words(s3))

# Approach 1: Built-in methods
def reverse_words(s):
    s = s.split()
    print(s)
    s = " ".join(reversed(s))
    return s

print(reverse_words(s0))

File: leetcode/medium/151.Reverse Words in a String/js/reverseWordsInAString.js
"use strict";
// Reverse Words in a String

// Approach 1: Built-in methods
var reverseWords = function(s) {
    // remove leading and trailing spaces
    s = s.trim();
    // split by spaces and reverse
    let words = s.split(/\s+/).reverse();
    // join the words with a space
    return words.join(" ");
}

// Testing
let s0 = "   the      sky is blue   ";
console.log(reverseWords(s0));

// Complexity Analysis
// Time complexity: O(N), where N is the number of characters in the input string.
// Space complexity: O(N), to store the result of split by spaces.